astropy/utils/diff.py

```diff
@@ -163,25 +163,49 @@ def where_not_allclose(a, b, rtol=1e-5, atol=1e-8):
     ----------
     a, b : array-like
         Input arrays to compare.
-
     rtol, atol : float
         Relative and absolute tolerances as accepted by
         :func:`numpy.allclose`.
+    return_maxdiff : bool
+        Return the maximum of absolute and relative differences.
 
     Returns
     -------
     idx : tuple of array
         Indices where the two arrays differ.
+    max_absolute : float
+        Maximum of absolute difference, returned if ``return_maxdiff=True``.
+    max_relative : float
+        Maximum of relative difference, returned if ``return_maxdiff=True``.
 
     """
     # Create fixed mask arrays to handle INF and NaN; currently INF and NaN
     # are handled as equivalent
-    if not np.all(np.isfinite(a)):
-        a = np.ma.fix_invalid(a).data
-    if not np.all(np.isfinite(b)):
-        b = np.ma.fix_invalid(b).data
+    a = np.ma.masked_invalid(a)
+    b = np.ma.masked_invalid(b)
+
+    absolute = np.ma.abs(b - a)
 
     if atol == 0.0 and rtol == 0.0:
         # Use a faster comparison for the most simple (and common) case
-        return np.where(a != b)
-    return np.where(np.abs(a - b) > (atol + rtol * np.abs(b)))
+        thresh = 0
+    else:
+        thresh = atol + rtol * np.abs(b)
+
+    # values invalid in only one of the two arrays should be reported
+    invalid = a.mask ^ b.mask
+    indices = np.where(invalid | (absolute.filled(0) > thresh))
+
+    if return_maxdiff:
+        if len(indices[0]) == 0:
```

Comment:
In the case where all elements are masked, this fails with
`ValueError: zero-size array to reduction operation maximum which has no identity`

You can avoid this by moving lines 204 and 205 before the if and adding a size condition as follows:
``` 
finites = ~absolute.mask
absolute = absolute.compressed()
if len(indices[0]) == 0 or absolute.size == 0:
```

Reply:
Thanks, done :+1: